### DISCUSSION TITLE
Can we save state machine JSON/XML format in NWB?

### DISCUSSION THREAD

**Discussion by tuanpham96** [NeurodataWithoutBorders/helpdesk#42] [General]
Is there a way to save an XML or JSON data in NWB? Are there certain NWB data types that already handle JSON/XML objects?

The reason being many behavior experiments are designed with the state machine framework. 

And it would be great just to save such set up in a certain format, instead of appending it as text inside `.description` and `.comments`. 

Maybe `ndx-beadl` has already touched upon some of this already but I'm unsure how to use it with files not from BEADL.

**Comment by CodyCBakerPhD** [comment]
Hello @tuanpham96

I agree this is something we can and should improve upon as we continue development on the NWB standard.

It heavily depends on the structure of what you're serializing in JSON, but generally speaking I'd say shape it into a DynamicTable instead of dumping the string in a free-text field.

Dumping it to a `description` is something we occasionally see users do; the only thing we don't recommend is repeating the JSON serialization over items within a table. We have a specific [Best Practice](https://nwbinspector.readthedocs.io/en/dev/best_practices/tables.html?highlight=JSON#table-values) against that.

If you're using Python, this process can be as simple as in [this example](https://stackoverflow.com/a/21266043/13616096) (assuming the string is able to be 'normalized') at which point we have [existing convenience functions](https://nwbinspector.readthedocs.io/en/dev/best_practices/tables.html?highlight=JSON#table-values) for automatically creating a DynamicTable from that DataFrame.

**Comment by CodyCBakerPhD** [comment]
As a long-term solution however, just making a separate table for the information loses the original linkage to the corresponding neurodata type so that's not really ideal...

A common occurrence of this from what I've seen is system, trial, or other processing-specific **configurations**, so I've wondered if it might be useful to add a general field to base data types (TimeSeries, TimeIntervals, etc.) called `config` that could itself be a DynamicTable or link to one and could serve as 'the place' where you store such machine-readable information... 

Any thoughts on that @rly @oruebel @bendichter?

**Comment by oruebel** [comment]
> Is there a way to save an XML or JSON data in NWB? Are there certain NWB data types that already handle JSON/XML objects?

The `ndx-beadl` extension has fields for `TaskProgram` for the actual program and `TaskSchema` for the schema. I.e,. you could store the JSON of the program in `TaskProgram` field and the JSONSchema in the `TaskSchema` field.

> Maybe `ndx-beadl` has already touched upon some of this already but I'm unsure how to use it with files not from BEADL.

I think `ndx-beadl` could indeed be a good solution for this. While the extension was designed with BEADL, it was designed with the goal to be generalizable to other state-machine models of behavior. Aside from the `BEADLTaskProgam` (which is just a special type of the `TaskProgram`), all other data types in the extension are general table-based structures based on the ``DynamicTable`` type. ``ndx-beadl`` defines tables to: 1) defines types of states, events, and actions, 2) acquisitions of when states, event, and actions occurred, as well as 3) a ``TrialsTable`` that extends ``TimeIntervals`` to group states, events, and actions by trials. 

If you have questions about using ``ndx-beadl`` I'd be happy to also jump on a Zoom.

**Reply by tuanpham96** [reply]
Thank you both for your suggestions! 

@oruebel I have quite a few questions about how to use `ndx-beadl` for non-BEADL programs yet designed with state-machine models. I'll send you an email about Zoom.

**Reply by tuanpham96** [reply]
I sent you an email to oruebel@lbl.gov via tuanhpham@brown.edu. Let me know if there's another email I should use.

**Reply by oruebel** [reply]
Sorry, for the delay. I received your email and will get back to you today.
